pkg/nodeagent/controller/healthcheck/kubelet.go

```diff
@@ -0,0 +1,271 @@
+}
+
+// HasLastInternalIP returns true if the node.InternalIP was stored
+// exported for testing
+func (k *KubeletHealthChecker) HasLastInternalIP() bool {
+	return k.lastInternalIP.IsValid()
+}
+
+// SetKubeletHealthEndpoint set the kubeletHealthEndpoint
+// exported for testing
+func (k *KubeletHealthChecker) SetKubeletHealthEndpoint(kubeletHealthEndpoint string) {
+	k.kubeletHealthEndpoint = kubeletHealthEndpoint
+}
+
+// Check performs the actual health check for the kubelet.
+func (k *KubeletHealthChecker) Check(ctx context.Context) error {
+	log := logf.FromContext(ctx).WithName("kubelet")
+	node := &corev1.Node{}
+	err := k.client.Get(ctx, types.NamespacedName{Name: k.nodeName}, node)
+	if err != nil {
+		return fmt.Errorf("unable to fetch node %w", err)
+	}
+
+	// This mimics the old behavior of the kubelet-health-monitor which checks for
+	// to many Ready->NoReady toggles in a certain time and reboots the node in such a case.
+	if k.nodeReady != isNodeReady(node) {
+		needsReboot := k.ToggleKubeletState()
+		if needsReboot {
+			log.Info("Kubelet toggled between Ready and NotReady too often, reboot the node now")
+			k.recorder.Eventf(node, corev1.EventTypeWarning, "kubelet", "toggled between Ready and NotReady more than %d times in a %s time window, reboot the node now to resolve this", maxToggles, toggleTimeSpan)
+			err = k.dbus.Reboot()
+			if err != nil {
+				return fmt.Errorf("unable to reboot node %w", err)
+			}
+		}
+	}
+	k.nodeReady = isNodeReady(node)
+
+	err = k.ensureNodeInternalIP(ctx, node)
+	if err != nil {
+		return err
+	}
+
+	client := &http.Client{
+		Timeout: 10 * time.Second,
+	}
+	request, err := http.NewRequestWithContext(ctx, http.MethodGet, k.kubeletHealthEndpoint, nil)
+	if err != nil {
+		log.Error(err, "Unable to fetch kubelet health")
+		return err
+	}
+	response, err := client.Do(request)
+	if err != nil {
+		log.Error(err, "Unable to fetch kubelet health with http get request")
+	}
+	if err == nil && response.StatusCode == http.StatusOK {
+		if k.firstFailure != nil {
+			log.Info("Kubelet is healthy again")
+			k.recorder.Event(node, corev1.EventTypeNormal, "kubelet", "healthy")
+			k.firstFailure = nil
+		}
+		log.Info("Kubelet is healthy again", "statusCode", response.StatusCode)
+		return nil
+	}
+	if k.firstFailure == nil {
+		now := k.Clock.Now()
+		k.firstFailure = &now
+		log.Error(err, "Kubelet is not healthy")
+		k.recorder.Event(node, corev1.EventTypeWarning, "kubelet", "unhealthy")
+	}
+
+	if time.Since(*k.firstFailure).Abs() < maxFailureDuration {
+		return nil
+	}
+
+	log.Error(err, "Kubelet is not healthy, restarting")
+	err = k.dbus.Restart(ctx, k.recorder, node, kubeletServiceName)
+	if err == nil {
+		k.firstFailure = nil
+	}
+	return err
+}
+
+// ensureNodeInternalIP mimics the old weird logic which restores the internalIP of the node
+// if this was lost for some reason.
+func (k *KubeletHealthChecker) ensureNodeInternalIP(ctx context.Context, node *corev1.Node) error {
```

Comment:
Is this still required?
This came from https://github.com/gardener/gardener/issues/283, https://github.com/gardener/gardener/commit/1311de43a1745cbc8cf65d57c72e9ed0a2c5e586#diff-738db1352694482843441061260a6f02 and https://github.com/gardener/gardener/pull/558, but it is not clear whether this is still relevant.

Reply:
@timuthy / @rfranzke As you were involved in the original logic, do you think that we still need it?

From my point of view, we can keep it, but it would be nice if it would be documented clearly in what scenario it is needed.

Reply:
Well, what should I say ... I don't know if and under which circumstances https://github.com/gardener/gardener/issues/283 is still relevant today 😄

Reply:
https://github.com/gardener/gardener/commit/1311de43a1745cbc8cf65d57c72e9ed0a2c5e586 mentions that the Addresses were lost if Kubelet ran into a `10s` timeout while calling the cloud provider.
I browsed through the current Kubelet code and it seems like the operation is retried implicitly but many areas have changed since 1.11 which was expected. Hence, I can't tell for sure if this workaround can be dropped.

We should open a follow-up issue and verify this separately.

> but it would be nice if it would be documented clearly in what scenario it is needed.

Absolutely! The current doc string isn't helpful, rather blaming ;)